### PR TITLE
Drop support for DBAL 2

### PR DESCRIPTION
**Pull request by dmaicher**
Closes https://github.com/doctrine/DoctrineBundle/issues/1550

### CODE REVIEW

File: Controller/ProfilerController.php [hunk 1]
```diff
@@ -128,17 +125,6 @@ private function explainSQLServerPlatform(Connection $connection, array $query):
 
         $stmt = $connection->executeQuery($sql, $params, $query['types']);
 
-        // DBAL 2.13 "forward compatibility" BC break handling
-        if ($stmt instanceof Result) {
```
Comment (dmaicher) [inline Controller/ProfilerController.php:132]: so I guess with DBAL 3 this never worked :thinking:

File: Controller/ProfilerController.php [hunk 2]
```diff
@@ -128,17 +125,6 @@ private function explainSQLServerPlatform(Connection $connection, array $query):
...
-            $stmt = $stmt->getIterator();
-        }
-
-        if (! $stmt instanceof PDOStatement) {
-            throw new LogicException('We need nextRowSet() functionality feature, which is not available with current DBAL driver');
```
Comment (dmaicher) [inline Controller/ProfilerController.php:137]: anyone knows why we need to call `nextRowSet` here?

File: Tests/ConnectionFactoryTest.php [hunk 3]
```diff
@@ -7,59 +7,19 @@
...
-use function defined;
-
-// Compatibility with DBAL < 3
-class_exists(\Doctrine\DBAL\Platforms\MySqlPlatform::class);
 
 class ConnectionFactoryTest extends TestCase
 {
     use VerifyDeprecations;
 
-    public function testContainer(): void
```
Comment (dmaicher) [inline Tests/ConnectionFactoryTest.php:32]: I don't understand what this test actually tests. That exception throwing logic all seems DBAL related.

reading https://github.com/doctrine/DoctrineBundle/issues/673 I don't think its relevant anymore.

File: composer.json [hunk 4]
```diff
@@ -29,10 +29,10 @@
     ],
     "homepage": "https://www.doctrine-project.org",
     "require": {
-        "php": "^7.1 || ^8.0",
+        "php": "^7.3 || ^8.0",
```
Comment (derrabus) [inline composer.json]: PHP 7.3 is supported by unmaintained DBAL 3 branches only. We should bump to PHP 7.4 directly.

See also the [statistics on Packagist](https://packagist.org/packages/doctrine/dbal/php-stats#3): PHP 7.3 accounts for only 1.8% of all DBAL 3 installations.

File: Command/DoctrineCommand.php [hunk 5]
```diff
@@ -52,36 +46,13 @@ protected function getEntityGenerator()
...
      * Get a doctrine entity manager by symfony name.
      *
-     * @param string   $name
-     * @param int|null $shardId
+     * @param string $name
      *
      * @return EntityManager
      */
-    protected function getEntityManager($name, $shardId = null)
+    protected function getEntityManager($name)
```
Comment (derrabus) [inline Command/DoctrineCommand.php]: Passing a shard ID to this method should raise an exception as before.

File: DependencyInjection/DoctrineExtension.php [hunk 6]
```diff
@@ -643,7 +536,7 @@ protected function ormLoad(array $config, ContainerBuilder $container)
         $container->registerForAutoconfiguration(AbstractIdGenerator::class)
             ->addTag(IdGeneratorPass::ID_GENERATOR_TAG);
 
-        if (method_exists($container, 'registerAttributeForAutoconfiguration')) {
+        if (PHP_VERSION_ID >= 80000 && method_exists(ContainerBuilder::class, 'registerAttributeForAutoconfiguration')) {
```
Comment (derrabus) [inline DependencyInjection/DoctrineExtension.php]: Why is this change necessary?

Comment (dmaicher) [reply DependencyInjection/DoctrineExtension.php]: Psalm was complaining with `RedundantCondition` :man_shrugging: I aligned this with https://github.com/doctrine/DoctrineBundle/blob/4bbe5880469d3cfab660f4dd887959b22bcf9dc4/DependencyInjection/DoctrineExtension.php#L162

Comment (derrabus) [reply DependencyInjection/DoctrineExtension.php]: That's a false positive. Psalm analyzes the code with Symfony 6 deps installed and this condition makes sure we don't run into errors with Symfony 4. We should rather suppress the error here.

Comment (dmaicher) [reply DependencyInjection/DoctrineExtension.php]: Ok I don't mind. Changed it back.

File: Controller/ProfilerController.php [hunk 7]
```diff
@@ -128,17 +125,6 @@ private function explainSQLServerPlatform(Connection $connection, array $query):
...
-        }
-
-        $stmt->nextRowset();
-
         return $stmt->fetchAll(PDO::FETCH_ASSOC);
```
Comment (derrabus) [inline Controller/ProfilerController.php]: This doesn't work with DBAL 3 either, does it. I'm tempted to replace this method with an exception and a call for contributors, tbh. 🙈

Comment (dmaicher) [reply Controller/ProfilerController.php]: yeah I have no easy way of testing this :cry: :see_no_evil: seems completely broken indeed

Comment (dmaicher) [reply Controller/ProfilerController.php]: Although maybe it works?

That method should still work: https://github.com/doctrine/dbal/blob/3.5.x/src/Result.php#L300

Comment (derrabus) [reply Controller/ProfilerController.php]: Yes, but it does not accept PDO constants anymore.

Comment (dmaicher) [reply Controller/ProfilerController.php]: Well... `FetchMode::ASSOCIATIVE = PDO::FETCH_ASSOC = 2` :see_no_evil:

Comment (derrabus) [reply Controller/ProfilerController.php]: What a lucky coincidence. 😁 

Shall we just switch to `fetchAllAssociative()` and cross fingers then?

File: .github/workflows/continuous-integration.yml [hunk 8]
```diff
@@ -21,9 +21,6 @@ jobs:
       fail-fast: false
       matrix:
         php-version:
-          - "7.1"
-          - "7.2"
-          - "7.3"
           - "7.4"
```
Comment (stof) [inline .github/workflows/continuous-integration.yml:24]: we should add jobs on PHP 8

File: .github/workflows/continuous-integration.yml [hunk 9]
```diff
@@ -81,7 +78,7 @@ jobs:
 
       - name: "Install PHP with PCOV"
         uses: "shivammathur/setup-php@v2"
-        if: "${{ matrix.php-version != '7.1' }}"
+        if: "${{ matrix.php-version != '7.4' }}"
```
Comment (stof) [inline .github/workflows/continuous-integration.yml]: This change looks wrong to me. We should keep using PCov on PHP 7.4 (removing the PHP 7.1 logic entirely)

File: Controller/ProfilerController.php (hunk 2 above)
Comment (stof) [reply Controller/ProfilerController.php:137]: I think it is related to the way SQL server returns the statistics

File: Resources/doc/configuration.rst [hunk 10]
```diff
@@ -562,32 +494,6 @@ Configuration Reference
                         <!-- sslcrl: The name of a file containing the SSL certificate revocation list (CRL) -->
                         <!-- pooled: True to use a pooled server with the oci8/pdo_oracle driver -->
                         <!-- MultipleActiveResultSets: Configuring MultipleActiveResultSets for the pdo_sqlsrv driver -->
```
Comment (stof) [inline Resources/doc/configuration.rst]: You should remove these comments too

File: .github/workflows/continuous-integration.yml [hunk 11]
```diff
@@ -81,7 +80,6 @@ jobs:
 
       - name: "Install PHP with PCOV"
         uses: "shivammathur/setup-php@v2"
-        if: "${{ matrix.php-version != '7.1' }}"
```
Comment (derrabus) [inline .github/workflows/continuous-integration.yml:84]: If we do that, the `"Install PHP with XDebug"` step does not make sense anymore. Let's remove it.

Comment (dmaicher) [reply .github/workflows/continuous-integration.yml:84]: done

File: Command/CreateDatabaseDoctrineCommand.php [hunk 12]
```diff
@@ -60,51 +56,17 @@ protected function execute(InputInterface $input, OutputInterface $output): int
...
-        }
-
         $hasPath = isset($params['path']);
         $name    = $hasPath ? $params['path'] : ($params['dbname'] ?? false);
         if (! $name) {
             throw new InvalidArgumentException("Connection does not contain a 'path' or 'dbname' parameter and cannot be created.");
         }
 
         // Need to get rid of _every_ occurrence of dbname from connection configuration and we have already extracted all relevant info from url
+        /** @psalm-suppress InvalidArrayOffset */
```
Comment (ostrolucky) [inline Command/CreateDatabaseDoctrineCommand.php:66]: What was the error here? Does it complain about `url` key? Because that looks correct to me,  `$connection->getParams()` cannot really return `url` key, can it? Same in DropDatabaseDoctrineCommand

File: Command/DoctrineCommand.php [hunk 13]
```diff
@@ -61,24 +56,8 @@ protected function getEntityManager($name, $shardId = null)
...
-                throw new LogicException(sprintf('Sharding is supported only in EntityManager of instance "%s".', EntityManagerInterface::class));
-            }
-
-            $connection = $manager->getConnection();
-            if (! $connection instanceof PoolingShardConnection) {
-                throw new LogicException(sprintf("Connection of EntityManager '%s' must implement shards configuration.", $name));
-            }
-
-            $connection->connect($shardId);
+        if ($shardId !== null) {
```
Comment (ostrolucky) [inline Command/DoctrineCommand.php:59]: I would just remove argument

File: Controller/ProfilerController.php (hunk 1 above)
Comment (ostrolucky) [reply Controller/ProfilerController.php:132]: Then let's just throw exception in this function to make it clear this is not supported for SQLSRV? Because I'm pretty sure it ccannot work without nextRowset()

File: Tests/DependencyInjection/DoctrineExtensionTest.php [hunk 14]
```diff
@@ -789,6 +785,9 @@ public function testAnnotationsBundleMappingDetection(): void
         ]);
     }
 
+    // Disabled to prevent changing the comment below to a single-line annotation
+    // phpcs:disable SlevomatCodingStandard.Commenting.RequireOneLineDocComment.MultiLineDocComment
```
Comment (ostrolucky) [inline Tests/DependencyInjection/DoctrineExtensionTest.php]: better to put exclude rule in phpcs.xml.dist, or to use if PHP_VERSION_ID + $this->skip, wouldn't it?

**Review by ostrolucky — COMMENTED** [review]
nice work, few minor stuff

File: Tests/DependencyInjection/DoctrineExtensionTest.php (hunk 14 above)
Comment (derrabus) [reply Tests/DependencyInjection/DoctrineExtensionTest.php]: PHPUnit 9.5 can handle single-line annotations. And if we bump to PHP 7.4, we don't need to care about earlier PHPUnit versions.

**Comment by derrabus** [comment]
In composer.json, we can drop older PHPUnit versions:

```JSON
"phpunit/phpunit": "^9.5.26 || ^10.0",
```

File: Command/DoctrineCommand.php (hunk 13 above)
Comment (dmaicher) [reply Command/DoctrineCommand.php:59]: At first I removed it but then I added it back. See https://github.com/doctrine/DoctrineBundle/pull/1570#discussion_r1010801285

File: Command/CreateDatabaseDoctrineCommand.php (hunk 12 above)
Comment (dmaicher) [reply Command/CreateDatabaseDoctrineCommand.php:66]: Yes exactly. The error was about the `url` key.

> $connection->getParams() cannot really return url key, can it?

Actually it can :cry: I tested it. If I configure a DBAL connection using `url` in the bundle config then that key is also present in connection params.

File: Controller/ProfilerController.php (hunk 1 above)
Comment (dmaicher) [reply Controller/ProfilerController.php:132]: removed + added an exception

File: Command/DoctrineCommand.php (hunk 13 above)
Comment (ostrolucky) [reply Command/DoctrineCommand.php:59]: alright then

**Review by derrabus — APPROVED** [review]

**Review by ostrolucky — APPROVED** [review]

**Comment by derrabus** [comment]
Thank you, @dmaicher!
